core/invoker/src/main/scala/whisk/core/invoker/ActionLogDriver.scala

```diff
@@ -0,0 +1,119 @@
+        val bytes = log.getBytes(StandardCharsets.UTF_8).take(maxBytes.toBytes.toInt)
+        LogLine(time, stream, new String(bytes, StandardCharsets.UTF_8))
+    }
+}
+
+protected[invoker] object LogLine extends DefaultJsonProtocol {
+    implicit val serdes = jsonFormat3(LogLine.apply)
+}
+
+protected[invoker] trait ActionLogDriver {
+
+    // The action proxies inserts this line in the logs at the end of each activation for stdout/stderr
+    protected val LOG_ACTIVATION_SENTINEL = "XXX_THE_END_OF_A_WHISK_ACTIVATION_XXX"
+
+    /**
+     * Given the JSON driver's raw output of a docker container, convert it into our own
+     * JSON format. If asked, check for sentinel markers (which are not included in the output).
+     *
+     * Only parses and returns so much logs to fit into the LogLimit passed.
+     *
+     * @param logMsgs raw String read from a JSON log-driver written file
+     * @param requireSentinel determines if the processor should wait for a sentinel to appear
+     * @param limit the limit to apply to the log size
+     *
+     * @return Tuple containing (isComplete, isTruncated, logs)
+     */
+    protected def processJsonDriverLogContents(logMsgs: String, requireSentinel: Boolean, limit: ByteSize)(
+        implicit transid: TransactionId, logging: Logging): (Boolean, Boolean, Vector[LogLine]) = {
+
+        if (logMsgs.nonEmpty) {
+            val records = logMsgs.split("\n").toStream flatMap {
+                line =>
+                    Try { line.parseJson.convertTo[LogLine] } match {
+                        case Success(t) => Some(t)
+
+                        case Failure(t) =>
+                            // Drop lines that did not parse to JSON objects.
+                            // However, should not happen since we are using the json log driver.
+                            logging.error(this, s"log line skipped/did not parse: $t")
+                            None
+                    }
+            }
+
+            val cumulativeSize = records.scanLeft(0.bytes) { (acc, current) =>
+                acc + {
+                    // do not count the sentinels
+                    if (current.log != LOG_ACTIVATION_SENTINEL) current.log.sizeInBytes else 0.B
+                }
+            }.tail
+
+            val logBuffer = records.zip(cumulativeSize).takeWhile(_._2 <= limit)
+            val isTruncated = logBuffer.size < records.size
+
+            if (requireSentinel) {
+                // there may be more than one sentinel in stdout/stderr (as logs may leak across activations
+                // if for example there log limit was exceeded in one activation and the container was reused
+                // to run the same action again (this is considered a feature - otherwise, must drain the logs
+                // or destroy the container as if it errored)
+                val (sentinels, regulars) = logBuffer.partition(_._1.log.trim == LOG_ACTIVATION_SENTINEL)
```

Comment:
Shall we do the `trim` in the jsonFormat of `LogLine`? It's used again in the format of the log and it'd be neat if you can just rely on it being trimmed instead of doing it just in case?